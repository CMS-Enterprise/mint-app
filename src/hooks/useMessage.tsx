import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState
} from 'react';
import { useLocation } from 'react-router-dom';

const MessageContext = createContext<
  | {
      message: string | React.ReactNode | undefined;
      showMessage: (message: string | React.ReactNode) => void;
      showMessageOnNextPage: (message: string | React.ReactNode) => void;
    }
  | undefined
>(undefined);

// MessageProvider manages the state necessary for child components to
// use the useMessage hook.
const MessageProvider = ({ children }: { children: ReactNode }) => {
  const [queuedMessage, setQueuedMessage] = useState<
    string | React.ReactNode
  >();
  const [message, setMessage] = useState<string | React.ReactNode>();
  const location = useLocation();

  const [lastPathname, setLastPathname] = useState(location.pathname);

  useEffect(() => {
    if (lastPathname !== location.pathname) {
      setMessage(queuedMessage);
      setQueuedMessage('');
      setLastPathname(location.pathname);
    }
  }, [message, queuedMessage, lastPathname, location.pathname]);

  return (
    <MessageContext.Provider
      value={{
        message,
        showMessage: setMessage,
        showMessageOnNextPage: setQueuedMessage
      }}
    >
      {children}
    </MessageContext.Provider>
  );
};

// useMessage is a hook that provides the ability to show notifications to the
// user. Components will use the hook to get access to two methods:
//
// showMessage: display a message to the user immediately. This is useful
// when an interaction won't cause an immediate redirect to a new page.
//
// showMessageOnNextPage: schedule a message to be shown on the next
// visited by the user. Pages are identified using their URL and therefore
// messages won't appear until the URL changes.
//
// Currently there is only support for showing a single message.
const useMessage = () => {
  const context = useContext(MessageContext);
  if (context === undefined) {
    throw new Error('useMessage must be used within a MessageProvider');
  }

  return context;
};

export default useMessage;
export { useMessage, MessageProvider };
