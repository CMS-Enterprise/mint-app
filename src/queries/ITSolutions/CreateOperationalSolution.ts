import { gql } from '@apollo/client';

export default gql`
  mutation CreateOperationalSolution(
    $operationalNeedID: UUID!
    $solutionType: OperationalSolutionKey
    $changes: OperationalSolutionChanges!
  ) {
    createOperationalSolution(
      operationalNeedID: $operationalNeedID
      solutionType: $solutionType
      changes: $changes
    ) {
      id
      nameOther
      needed
      key
    }
  }
`;
