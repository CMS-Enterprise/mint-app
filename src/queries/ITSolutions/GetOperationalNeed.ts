import { gql } from '@apollo/client';

export default gql`
  query GetOperationalNeed($id: UUID!, $includeNotNeeded: Boolean = true) {
    operationalNeed(id: $id) {
      id
      modelPlanID
      name
      key
      nameOther
      needed
      solutions(includeNotNeeded: $includeNotNeeded) {
        id
        isOther
        name
        key
        pocName
        pocEmail
        needed
        nameOther
        mustStartDts
        mustFinishDts
        status
      }
    }
  }
`;
