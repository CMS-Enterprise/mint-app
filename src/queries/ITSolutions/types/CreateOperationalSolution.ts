/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { OperationalSolutionKey, OperationalSolutionChanges } from "./../../../types/graphql-global-types";

// ====================================================
// GraphQL mutation operation: CreateOperationalSolution
// ====================================================

export interface CreateOperationalSolution_createOperationalSolution {
  __typename: "OperationalSolution";
  id: UUID;
}

export interface CreateOperationalSolution {
  createOperationalSolution: CreateOperationalSolution_createOperationalSolution;
}

export interface CreateOperationalSolutionVariables {
  operationalNeedID: UUID;
  solutionType?: OperationalSolutionKey | null;
  changes: OperationalSolutionChanges;
}
