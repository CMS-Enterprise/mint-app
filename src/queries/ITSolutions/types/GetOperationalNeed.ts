/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { OperationalNeedKey, OperationalSolutionKey, OpSolutionStatus } from "./../../../types/graphql-global-types";

// ====================================================
// GraphQL query operation: GetOperationalNeed
// ====================================================

export interface GetOperationalNeed_operationalNeed_solutions {
  __typename: "OperationalSolution";
  id: UUID;
  name: string | null;
  key: OperationalSolutionKey | null;
  pocName: string | null;
  pocEmail: string | null;
  needed: boolean | null;
  nameOther: string | null;
  isOther: boolean;
  otherHeader: string | null;
  mustStartDts: Time | null;
  mustFinishDts: Time | null;
  status: OpSolutionStatus;
}

export interface GetOperationalNeed_operationalNeed {
  __typename: "OperationalNeed";
  id: UUID;
  modelPlanID: UUID;
  name: string | null;
  key: OperationalNeedKey | null;
  nameOther: string | null;
  needed: boolean | null;
  solutions: GetOperationalNeed_operationalNeed_solutions[];
}

export interface GetOperationalNeed {
  operationalNeed: GetOperationalNeed_operationalNeed;
}

export interface GetOperationalNeedVariables {
  id: UUID;
  includeNotNeeded?: boolean | null;
}
