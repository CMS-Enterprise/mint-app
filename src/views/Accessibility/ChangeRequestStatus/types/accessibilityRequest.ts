/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { AccessibilityRequestStatus } from "./../../../../types/graphql-global-types";

// ====================================================
// GraphQL query operation: accessibilityRequest
// ====================================================

export interface accessibilityRequest_accessibilityRequest_statusRecord {
  __typename: "AccessibilityRequestStatusRecord";
  status: AccessibilityRequestStatus;
}

export interface accessibilityRequest_accessibilityRequest {
  __typename: "AccessibilityRequest";
  id: UUID;
  name: string;
  statusRecord: accessibilityRequest_accessibilityRequest_statusRecord;
}

export interface accessibilityRequest {
  accessibilityRequest: accessibilityRequest_accessibilityRequest | null;
}

export interface accessibilityRequestVariables {
  id: UUID;
}
