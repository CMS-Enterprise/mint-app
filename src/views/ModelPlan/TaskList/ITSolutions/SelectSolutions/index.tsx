/*
View for selecting/toggled 'needed' bool on possible solutions and custom solutions
Displays relevant operational need question and answers
*/

import React, { useContext, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useHistory, useLocation, useParams } from 'react-router-dom';
import { useMutation, useQuery } from '@apollo/client';
import {
  Button,
  CardGroup,
  Grid,
  IconArrowBack
} from '@trussworks/react-uswds';
import { Form, Formik, FormikProps } from 'formik';

import Breadcrumbs from 'components/Breadcrumbs';
import PageHeading from 'components/PageHeading';
import Alert from 'components/shared/Alert';
import { ErrorAlert, ErrorAlertMessage } from 'components/shared/ErrorAlert';
import useMessage from 'hooks/useMessage';
import CreateOperationalSolution from 'queries/ITSolutions/CreateOperationalSolution';
import GetOperationalNeed from 'queries/ITSolutions/GetOperationalNeed';
import {
  CreateOperationalSolution as CreateOperationalSolutionType,
  CreateOperationalSolutionVariables
} from 'queries/ITSolutions/types/CreateOperationalSolution';
import {
  GetOperationalNeed as GetOperationalNeedType,
  GetOperationalNeed_operationalNeed as GetOperationalNeedOperationalNeedType,
  GetOperationalNeedVariables
} from 'queries/ITSolutions/types/GetOperationalNeed';
import {
  UpdateOperationalSolution as UpdateOperationalSolutionType,
  UpdateOperationalSolutionVariables
} from 'queries/ITSolutions/types/UpdateOperationalSolution';
import UpdateOperationalSolution from 'queries/ITSolutions/UpdateOperationalSolution';
import { OperationalNeedKey } from 'types/graphql-global-types';
import flattenErrors from 'utils/flattenErrors';
import { ModelInfoContext } from 'views/ModelInfoWrapper';
import NotFound from 'views/NotFound';

import CheckboxCard from '../_components/CheckboxCard';
import ITSolutionsSidebar from '../_components/ITSolutionSidebar';
import NeedQuestionAndAnswer from '../_components/NeedQuestionAndAnswer';

// Passing in operationalNeed to Formik instead of array of solutions
// Fomik does not take an array structure
export const initialValues: GetOperationalNeedOperationalNeedType = {
  __typename: 'OperationalNeed',
  id: '',
  modelPlanID: '',
  name: '',
  key: OperationalNeedKey.ACQUIRE_AN_EVAL_CONT,
  nameOther: '',
  needed: false,
  solutions: []
};

type SelectSolutionsProps = {
  update?: boolean;
};

const SelectSolutions = ({ update }: SelectSolutionsProps) => {
  const { t } = useTranslation('itSolutions');
  const { t: h } = useTranslation('draftModelPlan');
  const { modelID, operationalNeedID } = useParams<{
    modelID: string;
    operationalNeedID: string;
  }>();

  const {
    state: { isCustomNeed }
  } = useLocation<{
    isCustomNeed?: boolean;
  }>();

  const history = useHistory();

  const { showMessageOnNextPage } = useMessage();

  // State management for mutation errors
  const [mutationError, setMutationError] = useState<boolean>(false);

  const formikRef = useRef<FormikProps<GetOperationalNeedOperationalNeedType>>(
    null
  );

  const { modelName } = useContext(ModelInfoContext);

  const { data, loading, error } = useQuery<
    GetOperationalNeedType,
    GetOperationalNeedVariables
  >(GetOperationalNeed, {
    variables: {
      id: operationalNeedID,
      includeNotNeeded: true
    }
  });

  const operationalNeed = data?.operationalNeed || initialValues;

  const [createSolution] = useMutation<
    CreateOperationalSolutionType,
    CreateOperationalSolutionVariables
  >(CreateOperationalSolution);

  const [updateSolution] = useMutation<
    UpdateOperationalSolutionType,
    UpdateOperationalSolutionVariables
  >(UpdateOperationalSolution);

  // Cycles and updates all solutions on a need
  const handleFormSubmit = async (
    formikValues: GetOperationalNeedOperationalNeedType
  ) => {
    const { solutions } = formikValues;

    const removedSolutions: string = checkRemovedSolutions(
      operationalNeed,
      formikValues
    );

    await Promise.all(
      solutions.map(solution => {
        // if solution id is all zeros, it needs to be created
        if (solution.id === '00000000-0000-0000-0000-000000000000') {
          return createSolution({
            variables: {
              operationalNeedID,
              solutionType: solution.key,
              changes: {
                needed: solution.needed || false
              }
            }
          });
        }

        // Otherwise, set the NEEDED bool for solution
        // Custom Solution will have a "nameOther", otherwise, it will be empty
        return updateSolution({
          variables: {
            id: solution.id,
            changes: {
              needed: solution.needed || false,
              nameOther: solution.nameOther || ''
            }
          }
        });
      })
    )
      .then(response => {
        const errors = response?.find(result => result?.errors);

        if (response && !errors) {
          if (
            formikRef?.current?.values.solutions.find(
              solution => solution.needed
            ) ||
            update
          ) {
            showMessageOnNextPage(removedSolutions);
            history.push(
              `/models/${modelID}/task-list/it-solutions/${operationalNeedID}/${
                update ? 'update-status' : 'solution-implementation-details'
              }`,
              {
                fromSolutionDetails: false,
                isCustomNeed
              }
            );
          } else {
            history.push(`/models/${modelID}/task-list/it-solutions`);
          }
        } else if (errors) {
          setMutationError(true);
        }
      })
      .catch(err => {
        // setMutationError(true);
      });
  };

  const breadcrumbs = [
    { text: h('home'), url: '/' },
    { text: h('tasklistBreadcrumb'), url: `/models/${modelID}/task-list/` },
    { text: t('breadcrumb'), url: `/models/${modelID}/task-list/it-solutions` },
    {
      text: t('solutionDetails'),
      url: `/models/${modelID}/task-list/it-solutions/${operationalNeed.id}/${operationalNeed.solutions[0]?.id}/solution-details`
    },
    { text: update ? t('updateStatus') : t('selectSolution') }
  ];

  if (error) {
    return <NotFound />;
  }

  return (
    <>
      <Breadcrumbs
        items={
          update
            ? breadcrumbs
            : breadcrumbs.filter(item => item.text !== t('solutionDetails'))
        }
      />

      {mutationError && (
        <Alert type="error" slim>
          {t('addError')}
        </Alert>
      )}

      <Grid row gap>
        <Grid tablet={{ col: 9 }}>
          <PageHeading className="margin-top-4 margin-bottom-2">
            {update ? t('updateSolutions') : t('selectSolution')}
          </PageHeading>

          <p
            className="margin-top-0 margin-bottom-1 font-body-lg"
            data-testid="model-plan-name"
          >
            {h('for')} {modelName}
          </p>

          <p className="line-height-body-4">{t('selectInfo')}</p>

          <Grid tablet={{ col: 8 }} className="margin-bottom-4">
            <NeedQuestionAndAnswer
              operationalNeedID={operationalNeedID}
              modelID={modelID}
            />
          </Grid>

          {update && (
            <Alert type="info" slim className="margin-y-2">
              {t('updateSolutionsInfo')}
            </Alert>
          )}

          <Grid row gap>
            <Grid tablet={{ col: 10 }}>
              <Formik
                initialValues={operationalNeed}
                onSubmit={values => {
                  handleFormSubmit(values);
                }}
                enableReinitialize
                innerRef={formikRef}
              >
                {(
                  formikProps: FormikProps<GetOperationalNeedOperationalNeedType>
                ) => {
                  const { errors, handleSubmit, values } = formikProps;

                  const allTheSolutions = values.solutions;
                  const commonSolutions = allTheSolutions.filter(
                    solution => solution.isOther === false
                  );
                  const otherSolutions = allTheSolutions.filter(
                    solution => solution.isOther === true
                  );
                  const flatErrors = flattenErrors(errors);

                  return (
                    <>
                      {Object.keys(errors).length > 0 && (
                        <ErrorAlert
                          testId="formik-validation-errors"
                          classNames="margin-top-3"
                          heading={h('checkAndFix')}
                        >
                          {Object.keys(flatErrors).map(key => {
                            return (
                              <ErrorAlertMessage
                                key={`Error.${key}`}
                                errorKey={key}
                                message={flatErrors[key]}
                              />
                            );
                          })}
                        </ErrorAlert>
                      )}

                      <Form
                        className="margin-top-2"
                        data-testid="it-tools-page-seven-form"
                        onSubmit={e => {
                          handleSubmit(e);
                        }}
                      >
                        <legend className="text-bold margin-bottom-2">
                          {t('chooseCommonSolution')}
                        </legend>

                        {!loading && (
                          <CardGroup>
                            {commonSolutions.map((solution: any) => (
                              <CheckboxCard
                                solution={solution}
                                index={allTheSolutions.findIndex(
                                  x => x.name === solution.name
                                )}
                                key={solution.nameOther || solution.name}
                              />
                            ))}
                          </CardGroup>
                        )}

                        {otherSolutions.length > 0 && (
                          <>
                            <legend className="text-bold margin-top-5 margin-bottom-2">
                              {t('chooseOtherSolution')}
                            </legend>
                            {!loading && (
                              <CardGroup>
                                {otherSolutions.map((solution: any) => (
                                  <CheckboxCard
                                    solution={solution}
                                    index={allTheSolutions.findIndex(
                                      x => x.name === solution.name
                                    )}
                                    key={solution.nameOther || solution.name}
                                  />
                                ))}
                              </CardGroup>
                            )}
                          </>
                        )}

                        <Button
                          type="button"
                          id="add-solution-not-listed"
                          className="usa-button usa-button--outline margin-top-2"
                          onClick={() => {
                            history.push(
                              `/models/${modelID}/task-list/it-solutions/${operationalNeedID}/add-solution`,
                              { isCustomNeed }
                            );
                          }}
                        >
                          {t('selectAnother')}
                        </Button>

                        <div className="margin-top-6 margin-bottom-3">
                          <Button
                            type="submit"
                            className="margin-bottom-1"
                            disabled={
                              allTheSolutions.filter(
                                solution => solution.needed
                              ).length === 0
                            }
                          >
                            {t('continue')}
                          </Button>
                        </div>
                        <Button
                          type="button"
                          className="usa-button usa-button--unstyled display-flex flex-align-center"
                          onClick={() =>
                            history.push(
                              `/models/${modelID}/task-list/it-solutions`
                            )
                          }
                        >
                          <IconArrowBack
                            className="margin-right-1"
                            aria-hidden
                          />
                          {update ? t('dontUpdate') : t('dontAdd')}
                        </Button>
                      </Form>
                    </>
                  );
                }}
              </Formik>
            </Grid>
          </Grid>
        </Grid>
        <Grid tablet={{ col: 3 }} className="padding-x-1">
          <ITSolutionsSidebar modelID={modelID} renderTextFor="solution" />
        </Grid>
      </Grid>
    </>
  );
};

const checkRemovedSolutions = (
  operationalNeed: GetOperationalNeedOperationalNeedType,
  updatedNeed: GetOperationalNeedOperationalNeedType
) => {
  const removedSolutions: string[] = [];

  updatedNeed.solutions.forEach(solution => {
    const originalNeedSolution = operationalNeed.solutions.find(
      originalSolution =>
        (originalSolution.nameOther &&
          originalSolution.nameOther === solution.nameOther) ||
        originalSolution.id === solution.id
    );

    if (originalNeedSolution?.needed === true && solution.needed === false) {
      removedSolutions.push(
        // One of these values will never be null - ts doesn't recognize this
        // @ts-ignore
        originalNeedSolution.nameOther || originalNeedSolution.name
      );
    }
  });

  return removedSolutions.join(',');
};

export default SelectSolutions;
