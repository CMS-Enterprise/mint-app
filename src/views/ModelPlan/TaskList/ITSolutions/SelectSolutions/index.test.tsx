import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import {
  render,
  waitFor,
  waitForElementToBeRemoved
} from '@testing-library/react';
import { GetOperationalNeedDocument } from 'gql/gen/graphql';
import { GetOperationalNeed_operationalNeed as GetOperationalNeedType } from 'gql/gen/types/GetOperationalNeed';

import {
  needQuestionAndAnswerMock,
  possibleSolutionsMock
} from 'data/mock/solutions';
import { MessageProvider } from 'hooks/useMessage';
import CreateOperationalSolution from 'queries/ITSolutions/CreateOperationalSolution';
import {
  OperationalNeedKey,
  OperationalSolutionKey,
  OpSolutionStatus
} from 'types/graphql-global-types';
import setup from 'utils/testing/setup';

import SelectSolutions, { findChangedSolution } from '.';

const modelID = 'ce3405a0-3399-4e3a-88d7-3cfc613d2905';
const operationalNeedID = '081cb879-bd6f-4ead-b9cb-3a299de76390';

const operationalNeed: GetOperationalNeedType = {
  __typename: 'OperationalNeed',
  id: operationalNeedID,
  modelPlanID: modelID,
  name: 'Obtain an application support contractor',
  key: OperationalNeedKey.APP_SUPPORT_CON,
  nameOther: null,
  needed: true,
  solutions: [
    {
      __typename: 'OperationalSolution',
      id: '00000000-0000-0000-0000-000000000000',
      name: 'Research, Measurement, Assessment, Design, and Analysis (RMADA)',
      pocEmail: '',
      key: OperationalSolutionKey.RMADA,
      mustStartDts: null,
      mustFinishDts: null,
      status: OpSolutionStatus.AT_RISK,
      isOther: false,
      isCommonSolution: true,
      otherHeader: null,
      needed: null,
      pocName: 'John Doe',
      nameOther: null
    }
  ]
};

const mocks = [
  {
    request: {
      query: GetOperationalNeedDocument,
      variables: {
        id: operationalNeedID,
        includeNotNeeded: true
      }
    },
    result: {
      data: {
        operationalNeed
      }
    }
  },
  {
    request: {
      query: CreateOperationalSolution,
      variables: {
        operationalNeedID,
        solutionType: 'RMADA',
        changes: { needed: true }
      }
    },
    result: {
      data: {
        createOperationalSolution: {
          id: operationalNeedID,
          nameOther: null,
          needed: true,
          key: OperationalSolutionKey.RMADA
        }
      }
    }
  },
  ...needQuestionAndAnswerMock,
  ...possibleSolutionsMock
];

describe('Operational Solutions NeedQuestionAndAnswer', () => {
  it('renders correctly', async () => {
    const { user, getByText, getByRole } = setup(
      <MemoryRouter
        initialEntries={[
          {
            pathname: `/models/${modelID}/task-list/it-solutions/${operationalNeedID}/select-solutions`
          }
        ]}
      >
        <Route path="/models/:modelID/task-list/it-solutions/:operationalNeedID/select-solutions">
          <MockedProvider mocks={mocks} addTypename={false}>
            <MessageProvider>
              <SelectSolutions />
            </MessageProvider>
          </MockedProvider>
        </Route>
      </MemoryRouter>
    );

    await waitFor(async () => {
      const checkbox = getByRole('checkbox', {
        name: /Select this solution/i
      });
      expect(checkbox).not.toBeChecked();
      await user.click(checkbox);
      expect(checkbox).toBeChecked();
    });

    await waitFor(() => {
      expect(
        getByText('Research, Measurement, Assessment, Design, and Analysis')
      ).toBeInTheDocument();
    });

    getByRole('button', { name: /Continue/i }).click();

    await waitFor(() => {
      expect(
        getByText('Research, Measurement, Assessment, Design, and Analysis')
      ).toBeInTheDocument();
    });
  });

  it('returns changes solution boolean', async () => {
    expect(
      findChangedSolution(
        operationalNeed.solutions,
        operationalNeed.solutions[0]
      )
    ).toEqual(false);

    expect(
      findChangedSolution(operationalNeed.solutions, {
        __typename: 'OperationalSolution',
        id: '00000000-0000-0000-0000-000000000000',
        name: 'Research, Measurement, Assessment, Design, and Analysis',
        pocEmail: '',
        key: OperationalSolutionKey.RMADA,
        mustStartDts: null,
        mustFinishDts: null,
        isOther: false,
        otherHeader: '',
        isCommonSolution: true,
        status: OpSolutionStatus.AT_RISK,
        needed: true,
        pocName: 'John Doe',
        nameOther: null
      })
    ).toEqual(true);
  });

  it('matches snapshot', async () => {
    const { asFragment, getByText, getByTestId } = render(
      <MemoryRouter
        initialEntries={[
          {
            pathname: `/models/${modelID}/task-list/it-solutions/${operationalNeedID}/select-solutions`
          }
        ]}
      >
        <Route path="/models/:modelID/task-list/it-solutions/:operationalNeedID/select-solutions">
          <MockedProvider mocks={mocks} addTypename={false}>
            <MessageProvider>
              <SelectSolutions />
            </MessageProvider>
          </MockedProvider>
        </Route>
      </MemoryRouter>
    );

    await waitForElementToBeRemoved(() => getByTestId('page-loading'));

    await waitFor(() => {
      expect(getByText('at.mint@oddball.io')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
