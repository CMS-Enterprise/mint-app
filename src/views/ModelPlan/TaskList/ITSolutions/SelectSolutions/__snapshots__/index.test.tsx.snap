// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IT Solutions NeedQuestionAndAnswer matches snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="Breadcrumbs"
    class="usa-breadcrumb usa-breadcrumb--wrap"
  >
    <ol
      class="usa-breadcrumb__list"
    >
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          class="usa-link usa-breadcrumb__link"
          href="/"
        >
          <span>
            Home
          </span>
        </a>
      </li>
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          class="usa-link usa-breadcrumb__link"
          href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/task-list/"
        >
          <span>
            Model Plan task list
          </span>
        </a>
      </li>
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          class="usa-link usa-breadcrumb__link"
          href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/task-list/it-solutions"
        >
          <span>
            IT solutions tracker
          </span>
        </a>
      </li>
      <li
        aria-current="page"
        class="usa-breadcrumb__list-item usa-current"
      >
        <span>
          Select a solution
        </span>
      </li>
    </ol>
  </nav>
  <div
    class="grid-row grid-gap"
    data-testid="grid"
  >
    <div
      class="tablet:grid-col-9"
      data-testid="grid"
    >
      <h1
        aria-live="polite"
        class="mint-h1 margin-top-4 margin-bottom-2"
        tabindex="-1"
      >
        Select a solution
      </h1>
      <p
        class="margin-top-0 margin-bottom-1 font-body-lg"
        data-testid="model-plan-name"
      >
        for 
      </p>
      <p
        class="line-height-body-4"
      >
        Select from common solutions to solve your operational need or add another solution not listed.
      </p>
      <div
        class="tablet:grid-col-8 margin-bottom-4"
        data-testid="grid"
      >
        <div
          class="grid-container padding-3 bg-base-lightest maxw-none"
          data-testid="gridContainer"
        >
          <div
            class="grid-row"
            data-testid="grid"
          >
            <div
              class="desktop:grid-col-12"
              data-testid="grid"
            >
              <p
                class="text-bold margin-y-1"
              >
                Operational need
              </p>
              <p
                class="margin-top-0 font-body-sm"
              />
            </div>
            <div
              class="desktop:grid-col-12"
              data-testid="grid"
            >
              <a
                class="usa-link"
                href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/task-list/it-solutions/update-need/"
              >
                Edit this operational need
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="grid-row grid-gap"
        data-testid="grid"
      >
        <div
          class="tablet:grid-col-10"
          data-testid="grid"
        >
          <form
            action="#"
            class="margin-top-2"
            data-testid="it-tools-page-seven-form"
          >
            <legend
              class="text-bold margin-bottom-2"
            >
              Choose from common solutions
            </legend>
            <ul
              class="usa-card-group"
              data-testid="CardGroup"
            >
              <div
                class="tablet:grid-col-6"
                data-testid="grid"
              >
                <li
                  class="usa-card"
                  data-testid="Card"
                >
                  <div
                    class="usa-card__container"
                  >
                    <div
                      class="solutions-checkbox__header padding-2"
                    >
                      <div
                        class="usa-checkbox"
                        data-testid="checkbox"
                      >
                        <input
                          class="usa-checkbox__input"
                          id="it-solutions-rmada"
                          name="solutions[0].needed"
                          testid="it-solutions-rmada"
                          type="checkbox"
                          value="false"
                        />
                        <label
                          class="usa-checkbox__label"
                          for="it-solutions-rmada"
                        >
                          Select a solution
                        </label>
                      </div>
                      <h3
                        class="margin-y-2"
                      >
                        Research, Measurement, Assessment, Design, and Analysis (RMADA)
                      </h3>
                      <div
                        class="margin-bottom-2 solutions-checkbox__body-text"
                      >
                        Short summary. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore aliqa...
                      </div>
                      <p
                        class="text-bold margin-bottom-0"
                      >
                        Point of contact
                      </p>
                      <p
                        class="margin-y-0"
                      >
                        John Doe
                      </p>
                      <a
                        aria-label="Send an email"
                        class="usa-link line-height-body-5"
                        href="mailto:"
                        target="_blank"
                      >
                        <div
                          class="margin-bottom-2"
                        />
                      </a>
                      <div
                        class="mint-divider"
                      />
                      <button
                        class="usa-button display-flex flex-align-center usa-button usa-button--unstyled margin-y-2"
                        data-testid="button"
                        type="button"
                      >
                        About this solution
                        <svg
                          class="usa-icon margin-left-1"
                          focusable="false"
                          height="1em"
                          role="img"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M0 0h24v24H0z"
                            fill="none"
                          />
                          <path
                            d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </li>
              </div>
            </ul>
            <button
              class="usa-button usa-button usa-button--outline margin-top-2"
              data-testid="button"
              id="add-solution-not-listed"
              type="button"
            >
              Select another solution not listed
            </button>
            <div
              class="margin-top-6 margin-bottom-3"
            >
              <button
                class="usa-button margin-bottom-1"
                data-testid="button"
                disabled=""
                type="submit"
              >
                Continue
              </button>
            </div>
            <button
              class="usa-button usa-button usa-button--unstyled display-flex flex-align-center"
              data-testid="button"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="usa-icon margin-right-1"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                />
              </svg>
              Donâ€™t add solutions and return to tracker
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="tablet:grid-col-3 padding-x-1"
      data-testid="grid"
    >
      <div
        class="border-top-05 border-primary-lighter padding-top-2 margin-top-4"
      >
        <div
          class="padding-2 bg-primary-lighter"
        >
          <p
            class="text-bold margin-top-0"
          >
            Need help choosing a solution?
          </p>
          <div
            class="display-flex"
            data-testid="ask-a-question"
          >
            <svg
              class="usa-icon text-primary margin-right-1"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-7 9h-2V5h2v6zm0 4h-2v-2h2v2z"
              />
            </svg>
            <button
              class="usa-button usa-button--unstyled"
              data-testid="ask-a-question-button"
              type="button"
            >
              Ask a question
            </button>
          </div>
        </div>
      </div>
      <div
        class="margin-top-4"
      >
        <p
          class="text-bold margin-bottom-2"
        >
          Helpful links
        </p>
        <button
          class="usa-button usa-button usa-button--unstyled line-height-body-5"
          data-testid="button"
          type="button"
        >
          Available operational solutions (opens in a new tab)
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
